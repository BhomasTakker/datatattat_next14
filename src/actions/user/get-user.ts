"use server";

import { options } from "@/app/api/auth/[...nextauth]/options";
import { getLeanUserById } from "@/lib/mongo/actions/user/user";
import { initialiseServices } from "@/lib/services/intialise-services";
import { Session } from "@/types/auth/session";
import { IUser } from "@/types/user";
import { getServerSession } from "next-auth";

export async function getSessionUser() {
	const session = (await getServerSession(options)) as Session;
	if (!session) {
		// throw new Error("No session found");
		return null;
	}

	return session.user;
}

export async function getUserFromSessionId(id: string): Promise<IUser> {
	await initialiseServices();
	const user = await getLeanUserById(id);

	if (!user) {
		throw new Error("User not found");
		//redirect?
	}

	return user;
}

// getUser
export async function getUser(): Promise<IUser | null> {
	// try catch / if no session / if no user etc
	const sessionUser = await getSessionUser();
	if (!sessionUser) {
		return Promise.resolve(null);
	}
	return getUserFromSessionId(sessionUser.user_id);
}
