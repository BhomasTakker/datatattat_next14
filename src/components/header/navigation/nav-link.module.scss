@use "../../../scss/abstracts/colours/" as colours;
@use "../../../scss/abstracts/units/" as *;

.root {
	border-radius: 10%;
	padding: $spacingLg $spacingXs;

	text-transform: uppercase;
	white-space: nowrap;
	p {
		color: colours.$textColourDark;
	}

	&:hover,
	&:focus-visible {
		background-color: colours.$shadowLight0;
	}
}
