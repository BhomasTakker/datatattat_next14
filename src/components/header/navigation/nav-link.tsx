import Link from "next/link";

import styles from "./nav-link.module.scss";

export type NavLinkData = {
	route: string;
	label: string;
	key?: string;
};

type NavLinkProps = { isEdit: boolean } & NavLinkData;

export const NavLink = ({ route, label, isEdit }: NavLinkProps) => {
	const routeLink = isEdit ? `/edit?route=${route}` : route;
	// if edit...
	return (
		<Link className={`${styles.root}`} href={routeLink}>
			<p>{label}</p>
		</Link>
	);
};
