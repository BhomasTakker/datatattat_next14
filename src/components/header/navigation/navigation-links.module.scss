@use "../../../scss/abstracts/colours/" as colours;

.root {
	display: flex;
	justify-content: center;
	gap: 1rem;
	padding: 1rem 0;

	border-bottom: 1px solid colours.$shadowLight2;
}
