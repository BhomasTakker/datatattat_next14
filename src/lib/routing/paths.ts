export const signIn = (callback: string) => {
	return `/api/auth/signin?callbackUrl=${callback}`;
};

export const signOut = (callback: string) => {
	return `/api/auth/signout?callbackUrl=${callback}`;
};

export const home = () => {
	return "/";
};

export const edit = () => {
	return "/edit";
};

export const user = (username: string) => {
	return `/user/${username}`;
};

// rem these
export const member = () => {
	return "/Member";
};

export const clientMember = () => {
	return "/ClientMember";
};

// Should export as an object perhaps
