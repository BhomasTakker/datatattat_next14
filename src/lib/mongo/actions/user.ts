import { User } from "@/models/User";
import { IUser } from "@/types/user";

export const getUserBySignInEmail = async (email: string) => {
	console.log("getUserBySignInEmail User.findOne", email);
	const user = await User.findOne({ signin_email: email });
	console.log("getUserBySignInEmail User.findOne", user);
	return user;
	return await User.findOne({ signin_email: email });
};

export const getUserById = async (id: string) => {
	return (await User.findOne({ _id: id })) as IUser;
};

export const getUserByUsername = async (username: string) => {
	const decodedUsername = decodeURI(username);
	return (await User.findOne({ username: decodedUsername })) as IUser;
};

export const getAllUserByUsername = async (username: string) => {
	const decodedUsername = decodeURI(username);
	return (await User.find({ username: decodedUsername })) as IUser[];
};

export const createNewUser = async (user: Omit<IUser, "_id">) => {
	try {
		const newUser = new User(user);
		return { message: "User Created", user: await newUser.save() };
	} catch (err) {
		console.error(err);
		return { message: "Error creating user" };
	}
};

export const updateUser = async (id: string, user: Partial<IUser>) => {
	try {
		// create if doesn't exist. Perhaps should not be.
		const updatedUser = await User.findOneAndUpdate({ _id: id }, user, {
			new: true,
		});
		return { message: "User Updated" };
	} catch (err) {
		console.error(err);
		return { message: "Error updating user" };
	}
};
