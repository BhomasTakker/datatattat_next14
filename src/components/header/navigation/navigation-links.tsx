import React from "react";
import { NavLink, NavLinkData } from "./nav-link";
import styles from "./navigation-links.module.scss";

type NavigationProps = {
	navLinks: NavLinkData[];
	edit: boolean;
};

export const NavigationLinks = ({ navLinks, edit }: NavigationProps) => {
	const drawNavLinks = (links: NavLinkData[]) => {
		return links.map((link) => {
			const element = (
				<NavLink
					route={link.route}
					label={link.label}
					key={link.label}
					isEdit={edit}
				></NavLink>
			);
			return element;
		});
	};

	// Double div so we can center the nav
	return <div className={styles.root}>{drawNavLinks(navLinks)}</div>;
};
