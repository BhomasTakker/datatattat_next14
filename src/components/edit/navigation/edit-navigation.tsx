import Link from "next/link";
import styles from "./edit-navigation.module.scss";
import { NavigationMenu } from "@/components/header/navigation/navigation-menu";
import { getMainHeader, getSubHeaders } from "@/actions/header/get-header";
import { SubHeaders } from "@/components/header/sub-headers";
import { PATHS } from "@/lib/routing/paths";

type EditNavigationProps = {
	route: string;
	isAdminEdit?: boolean;
};

export const EditNavigation = async ({
	route,
	isAdminEdit,
}: EditNavigationProps) => {
	const mainHeader = await getMainHeader();
	const subHeaders = await getSubHeaders(route);

	// Get this from somewhere
	const homeHref = isAdminEdit ? PATHS.admin() : PATHS.edit();
	const homeLabel = isAdminEdit ? "Home" : `User Home`;

	const routePrefix = isAdminEdit ? PATHS.admin() : PATHS.edit();
	return (
		<section className={styles.header}>
			<Link href={homeHref}>{homeLabel}</Link>
			{/* BACK BUTTON / More up a level button */}
			{isAdminEdit ? (
				<NavigationMenu items={mainHeader.nav} prefix={routePrefix} />
			) : null}
			{subHeaders && subHeaders.length > 0 ? (
				<SubHeaders headersArray={subHeaders} prefix={routePrefix} />
			) : null}
		</section>
	);
};
