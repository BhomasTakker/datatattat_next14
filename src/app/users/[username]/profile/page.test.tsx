import React from "react";
import { render, screen, waitFor } from "@testing-library/react";
import UserProfile from "./page";
import { redirect } from "next/navigation";
import { PATHS } from "@/lib/routing/paths";
const mockGetUserByUsername =
	require("../../../../lib/mongo/actions/user").getUserByUsername;
const mockGetUser = require("../../../../actions/user/get-user").getUser;

// Mock dependencies
jest.mock("../../../../lib/mongo/actions/user", () => ({
	getUserByUsername: jest.fn(),
}));
jest.mock("../../../../actions/user/get-user", () => ({
	getUser: jest.fn(),
}));
jest.mock("../../../../lib/mongo/db", () => ({
	connectToMongoDB: jest.fn(),
}));
jest.mock("../../../../components/profile/profile-page", () => ({
	ProfilePage: ({ user }: any) => (
		<div data-testid="profile-page">{user.username}</div>
	),
}));
jest.mock("../../../../components/profile/profile-edit-page", () => ({
	ProfileEditPage: () => <div data-testid="profile-edit-page">Edit Page</div>,
}));
jest.mock("../../../../components/header/main-header", () => ({
	MainHeader: ({ route }: any) => (
		<div data-testid="main-header">{route.join("/")}</div>
	),
}));
jest.mock("next/navigation", () => ({
	redirect: jest.fn(),
}));

describe("UserProfile page", () => {
	const username = "testuser";
	const params = Promise.resolve({ username });

	beforeEach(() => {
		jest.clearAllMocks();
	});

	it("renders ProfileEditPage if session user matches profile user", async () => {
		const user = { _id: "123", username, avatar: "avatar.png" };
		mockGetUser.mockResolvedValue(user);
		mockGetUserByUsername.mockResolvedValue(user);

		const { container } = render(await UserProfile({ params }));

		expect(container).toMatchSnapshot();
		expect(screen.getByTestId("main-header")).toHaveTextContent(
			`users/${username}`
		);
		expect(screen.getByTestId("profile-edit-page")).toBeInTheDocument();
	});

	it("renders ProfilePage if session user does not match profile user", async () => {
		const sessionUser = { _id: "123", username: "otheruser" };
		const profileUser = { _id: "456", username, avatar: "avatar.png" };
		mockGetUser.mockResolvedValue(sessionUser);
		mockGetUserByUsername.mockResolvedValue(profileUser);

		const { container } = render(await UserProfile({ params }));

		expect(container).toMatchSnapshot();
		expect(screen.getByTestId("main-header")).toHaveTextContent(
			`users/${username}`
		);
		expect(screen.getByTestId("profile-page")).toHaveTextContent(username);
	});

	it("redirects to home if profile user does not exist", async () => {
		mockGetUser.mockResolvedValue({ _id: "123", username: "otheruser" });
		mockGetUserByUsername.mockResolvedValue(null);

		render(await UserProfile({ params }));

		expect(redirect).toHaveBeenCalledWith(PATHS.home());
	});
});
