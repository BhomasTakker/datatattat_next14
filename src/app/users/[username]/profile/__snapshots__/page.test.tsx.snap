// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`UserProfile page renders ProfileEditPage if session user matches profile user 1`] = `
<div>
  <div
    data-testid="main-header"
  >
    users/testuser
  </div>
  <div
    class="page"
  >
    <div
      data-testid="profile-edit-page"
    >
      Edit Page
    </div>
  </div>
</div>
`;

exports[`UserProfile page renders ProfilePage if session user does not match profile user 1`] = `
<div>
  <div
    data-testid="main-header"
  >
    users/testuser
  </div>
  <div
    class="page"
  >
    <div
      data-testid="profile-page"
    >
      testuser
    </div>
  </div>
</div>
`;
